### PR TITLE
[wip] CRM-20158, added code for storing card type and last 4 digit card num…

### PR DESCRIPTION
**Pull request by pradpnayak**
…ber in civicrm_financial_trxn table

----------------------------------------
* CRM-20158: Store card type and last 4 digits of credit card
  https://issues.civicrm.org/jira/browse/CRM-20158

### CODE REVIEW

**Comment by jackrabbithanna** [comment]
@KarinG @eileenmcnaughton 
Does anyone know if storing the last 4 digits of a CC number affects PCI compliance requirements or other security risks?

**Comment by xurizaemon** [comment]
> Does anyone know if storing the last 4 digits of a CC number affects PCI compliance requirements or other security risks?

This (and date storage) is queried frequently, it really should be in docs or wiki somewhere. Date and last four digits are not considered "sensitive" data.

* https://forum.civicrm.org/index.php?topic=34458.0
* http://security.stackexchange.com/questions/143642/pci-dss-v3-card-holder-data-storage
* http://security.stackexchange.com/questions/19860/minimum-requirements-for-storing-last-4-digits-of-credit-card-number

The actual PCI docs are the canonical reference tho, since standards can change. Currently v2: https://www.pcisecuritystandards.org/documents/pci_dss_v2.pdf

**Comment by KarinG** [comment]
Here's a current reference from the PCI Security Standards Council:
https://www.pcisecuritystandards.org/pdfs/pci_fs_data_storage.pdf

**On page 2:** under Technical Guidelines for Protecting Stored Payment Data:
• Truncation – removing a data segment, such as showing only the last four digits.

**Comment by eileenmcnaughton** [comment]
Whoops this PR has changed since I first looked - my review was based on adding the field to the table - it's not valid now.

File: CRM/Contribute/BAO/Query.php [hunk 1]
```diff
@@ -149,7 +150,11 @@ public static function where(&$query) {
       if (empty($query->_params[$id][0])) {
         continue;
       }
-      if (substr($query->_params[$id][0], 0, 13) == 'contribution_' || substr($query->_params[$id][0], 0, 10) == 'financial_'  || substr($query->_params[$id][0], 0, 8) == 'payment_') {
+      if (substr($query->_params[$id][0], 0, 13) == 'contribution_'
+        || substr($query->_params[$id][0], 0, 10) == 'financial_'
+        || substr($query->_params[$id][0], 0, 8) == 'payment_'
+        || $query->_params[$id][0] == 'card_type'
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: Let's keep the search changes in a separate PR - there are other Prs in the queue that alter all this logic

File: CRM/Contribute/Form/PaymentDetails.php [hunk 2]
```diff
@@ -0,0 +1,184 @@
+<?php
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/PaymentDetails.php]: how does this relate to adding the field?

**Comment by eileenmcnaughton** [comment]
OK - I took a quick look at this but it's really broad ranging & there are a lot of changes in here that would take a lot of work to review - I think resolving #9876 is a good idea - that can be the first focus

File: CRM/Contribute/Form/AdditionalPayment.php [hunk 3]
```diff
@@ -387,13 +400,7 @@ public function postProcess() {
...
     $contributionStatusID = CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Contribution', $this->_contributionId, 'contribution_status_id');
     if ($contributionStatuses[$contributionStatusID] == 'Pending') {
-      civicrm_api3('Contribution', 'create',
-        array(
-          'id' => $this->_contributionId,
-          'contribution_status_id' => array_search('Partially paid', $contributionStatuses),
-          'is_pay_later' => 0,
-        )
-      );
+      CRM_Core_DAO::setFieldValue('CRM_Contribute_DAO_Contribution', $this->_contributionId, 'contribution_status_id', array_search('Partially paid', $contributionStatuses));
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/AdditionalPayment.php]: This doesn't seem like the direction we are trying to go in

**Comment by eileenmcnaughton** [comment]
OK - there is a tonne of stuff in here that does not really seem related to the description, so we need to break this down.

We have added the fields to the civicrm_financial_trxn table. I feel like the next step is to define how they are set and updated through the api.

Let's assume we are working with the flow of creating a pending contribution, making a payment and then updating the contribution with the outcome, the standard core flow.

When the pending contribution is created an initial financial_trxn record is created. So, it would seem the  contribution.create api / BAO create function should support adding the card_type and pan_truncation for new payments. I'm not clear whether they should for payment updates? Potentially there could be more than one payment and hence it would be ambiguous. I also think it is clear cut that the completetransaction api should support card_type & pan_truncation fields.

So,  2 things that are clear cut in my mind and could be merged with an associated unit test would be updating the api for contribution to work with the the 2 new fields on create (but I still need to understand update better) and making the completetransaction api support it. Note that the code changes in both cases should be primarily in the BAO but the tests should wrap around the api.

**Comment by eileenmcnaughton** [comment]
I took another look but there is so much going on in here & a lot of it seems unrelated - ie. there is a whole additional form class in there, and the help text suggests the PAN should be entered by ? at some point, I'm not going to be able to review this as such a bit PR.

I still think figuring out how the api puts the data into the new fields is the best first step

**Comment by KarinG** [comment]
It should not require +710/-304 touching 23 files - to accomplish: "Added code for storing card type and last 4 digit card number"; that's a serious influx of code -> LexIm opportunity?

This is about handling credit cards - we can't get this wrong. 

@JoeMurray - what is this?

```
 $this->add('text', 'credit_card_number', ts('Card Number'), array(
      'size' => 5,
      'maxlength' => 4,
      'autocomplete' => 'off',
    ));
```

**Comment by eileenmcnaughton** [comment]
@KarinG I think that there is actually some good refactoring in there too - I can see signs of duplicate code being removed & consolidated. I'm just struggling to make sense of so many changes

File: CRM/Event/Form/Registration/Confirm.php [hunk 4]
```diff
@@ -1032,7 +1034,7 @@ public static function processContribution(
       $contribParams['address_id'] = CRM_Contribute_BAO_Contribution::createAddress($params, $form->_bltID);
     }
 
-    $contribParams['payment_processor'] = CRM_Utils_Array::value('payment_processor', $params);
+    $contribParams['payment_processor'] = CRM_Utils_Array::value('payment_processor_id', $params);
```
Comment (eileenmcnaughton) [inline CRM/Event/Form/Registration/Confirm.php]: minor point - if it's an id can the parameter be called payment_proceessor_id?

File: CRM/Member/Form/Membership.php [hunk 5]
```diff
@@ -275,10 +275,6 @@ public function setDefaultValues() {
 
     //setting default join date and receive date
     list($now, $currentTime) = CRM_Utils_Date::setDateDefaults();
-    if ($this->_action == CRM_Core_Action::ADD) {
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: See this confuses me - how does this fit in?

File: CRM/Member/Form/Membership.php [hunk 6]
```diff
@@ -338,10 +334,7 @@ public function setDefaultValues() {
         $defaults['source'] = $defaults['membership_source'];
       }
     }
-    //CRM-13420
-    if (empty($defaults['payment_instrument_id'])) {
-      $defaults['payment_instrument_id'] = key(CRM_Core_OptionGroup::values('payment_instrument', FALSE, FALSE, FALSE, 'AND is_default = 1'));
-    }
+    CRM_Contribute_Form_PaymentDetails::setDefaultValues($defaults, NULL);
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: Ah, ok - it's an extraction, which is a good thing, just hard to review the extraction with so many other chagnes

**Comment by JoeMurray** [comment]
@pradpnayak can you review each of the comments here, and either accept and act on suggestion eg by breaking out search stuff into a different PR, or answering questions eg @KarinG re: what is this on credit card widget with size 5 maxlength 4

**Comment by monishdeb** [comment]
I am marking this PR with WIP as the PR gone stale and needs works as mentioned by @eileenmcnaughton

**Comment by pradpnayak** [comment]
Submitted PR's in chunk
